content/blog/2021/10/02/MonitorTheSolver.adoc

```diff
@@ -0,0 +1,215 @@
+= Monitor the solver
+Christopher-Chianelli
+2021-10-02
+:page-interpolate: true
+:jbake-type: post
+:jbake-tags: monitoring, production
+
+It's 11pm on Friday evening. Everything was working fine -- until now.
+Suddenly services are failing left and right, and your boss want to know why.
+One tool in achieving a diagnosis is to use a monitoring system. Let's add monitoring to OptaPlanner and see if we can diagnose this issue.
+
+== What is monitoring?
+
+_Monitoring_ is observing the quality of a service over time.
+Mointoring is similar to logging, except its output is more easily analyzed by a machine.
+The software being monitored output _metrics_, a numerical measurement of some aspect of the software.
+The metrics then get recorded in a monitoring system, where it can be graphed, trigger alerts, and be correlated with events.
+
+For instance, you can monitor a web service.
+Its metrics could include the number of hits to a particular endpoint, time to respond to a request, and the number of failures.
+If the number of hits suddenly increase, it could correspond to the Slashdot effect or a potential Denial of Service attack.
+If the average time to respond to a request is increasing because there are not enough servers to fulfill requests, you might want to acquire additional servers.
+Additionally, monitoring allow you to see the effect a change has on the entire system.
+If you notice the time to respond to a request with the same load as before decreases after a change was pushed, you can conclude the change improved performance.
+
```

Comment:
Instead of explaining what monitoring can do for a REST service, can we explain what it does for the OptaPlanner?

Some advantages I see:
- How many solvers ran last night? How many are running now? Is that why my server room is so hot in the middle of the night?
- How long did they ran? If we double the number of datasets, solvers or solving time, will we get everything solved by 7 AM every day?
- Did any run abnormally long or surprisingly short? Any anomalies we should be investigating?
- Did any throw errors?